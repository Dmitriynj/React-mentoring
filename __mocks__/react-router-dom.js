/* eslint-disable no-underscore-dangle */

const originalModule = jest.requireActual('react-router-dom');
const mockedModule = jest.createMockFromModule('react-router-dom');

const reactRouterDom = {
  ...mockedModule,
  ...originalModule,
};

reactRouterDom.__setSpy = (name, value) => {
  reactRouterDom[name] = value;
};

module.exports = reactRouterDom;
