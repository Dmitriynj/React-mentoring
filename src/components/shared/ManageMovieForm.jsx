import React from 'react';
import PropTypes from 'prop-types';
import { useFormik } from 'formik';
import * as yup from 'yup';
import { Button, MenuItem } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { FormField } from '../FormField';

const GENRES = [
  'Drama',
  'Romance',
  'Animation',
  'Adventure',
  'Action',
  'Family',
  'Fantasy',
  'Comedy',
  'Science Fiction',
  'Documentary',
  'War',
  'Thriller',
];

const useStyles = makeStyles(() => ({
  genreItem: {
    '&.Mui-selected': {
      backgroundColor: '#f65261',
    },
    '&.Mui-selected:hover': {
      backgroundColor: '#f50057',
    },
  },
}));

const validationSchema = yup.object({
  title: yup.string('Enter title').required('Title is required'),
  release_date: yup.date('Enter release date').required('Release date is required'),
  poster_path: yup.string('Enter poster path').required('Poster path is required'),
  genres: yup.array().of(yup.mixed().oneOf(GENRES)).min(1, 'Add genre'),
  overview: yup.string('Enter overview').required('Overview is required'),
  runtime: yup.number('Enter runtime').required('Runtime is required'),
});

const ManageMovieForm = ({ onConfirm, initialMovie }) => {
  const classes = useStyles();

  const formik = useFormik({
    initialValues: initialMovie,
    validationSchema,
    onSubmit: (values) => {
      console.log('Form submitted', values);
      onConfirm(values);
    },
  });

  return (
    <form onSubmit={formik.handleSubmit}>
      <FormField
        alias="title"
        type="text"
        label="Title"
        value={formik.values.title || ''}
        onChange={formik.handleChange}
        touched={formik.touched.title}
        errorMsg={formik.errors.title}
      />
      <FormField
        alias="release_date"
        type="date"
        label="Release date"
        value={formik.values.release_date || ''}
        onChange={formik.handleChange}
        touched={formik.touched.release_date}
        errorMsg={formik.errors.release_date}
      />
      <FormField
        alias="poster_path"
        type="text"
        label="Poster path"
        value={formik.values.poster_path || ''}
        onChange={formik.handleChange}
        touched={formik.touched.poster_path}
        errorMsg={formik.errors.poster_path}
      />
      <FormField
        alias="genres"
        type="genres"
        label="Genres"
        value={formik.values.genres || ''}
        onChange={formik.handleChange}
        touched={formik.touched.genres}
        errorMsg={formik.errors.genres}
        select
        SelectProps={{
          multiple: true,
          MenuProps: {
            PaperProps: {
              style: {
                backgroundColor: '#232323',
                color: 'white',
              },
            },
          },
        }}
      >
        {GENRES.map((option) => (
          <MenuItem key={option} value={option} className={classes.genreItem}>
            {option}
          </MenuItem>
        ))}
      </FormField>
      <FormField
        alias="runtime"
        type="number"
        label="Runtime"
        value={formik.values.runtime}
        onChange={formik.handleChange}
        touched={formik.touched.runtime}
        errorMsg={formik.errors.runtime}
      />
      <FormField
        alias="overview"
        type="text"
        label="Overview"
        value={formik.values.overview || ''}
        onChange={formik.handleChange}
        touched={formik.touched.overview}
        errorMsg={formik.errors.overview}
      />
      <Button color="secondary" variant="contained" fullWidth type="submit">
        Submit
      </Button>
    </form>
  );
};

ManageMovieForm.propTypes = {
  onConfirm: PropTypes.func.isRequired,
  initialMovie: PropTypes.shape({
    title: PropTypes.string,
    release_date: PropTypes.string,
    poster_path: PropTypes.string,
    genres: PropTypes.array,
    runtime: PropTypes.number,
    overview: PropTypes.string,
  }),
};
ManageMovieForm.defaultProps = {
  initialMovie: {
    title: '',
    release_date: '',
    poster_path: '',
    genres: [],
    runtime: 0,
    overview: '',
  },
};

export { ManageMovieForm };
