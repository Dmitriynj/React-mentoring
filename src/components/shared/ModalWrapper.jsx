import React from 'react';
import PropTypes from 'prop-types';
import { Button, Modal, Typography } from '@material-ui/core';
import CloseIcon from '@material-ui/icons/Close';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles((theme) => ({
  paper: {
    position: 'absolute',
    width: 300,
    backgroundColor: '#232323',
    padding: theme.spacing(2, 4, 3),
    outline: 'none',
    border: 'none',
    borderRadius: 4,
  },
  root: {
    display: 'flex',
    flexDirection: 'column',
    color: '#ececec',
  },
  closeButton: {
    color: 'inherit',
    marginLeft: 'auto',
  },
}));

const ModalWrapper = ({ children, title, open, closeModal }) => {
  const classes = useStyles();
  const [modalStyle] = React.useState({
    top: `${50}%`,
    left: `${50}%`,
    transform: `translate(-${50}%, -${50}%)`,
  });

  return (
    <Modal open={open} onClose={closeModal}>
      <div style={modalStyle} className={classes.paper}>
        <div className={classes.root}>
          <Button className={classes.closeButton} onClick={closeModal}>
            <CloseIcon />
          </Button>
          <Typography variant="h4">{title}</Typography>
          {children}
        </div>
      </div>
    </Modal>
  );
};
ModalWrapper.propTypes = {
  children: PropTypes.element.isRequired,
  title: PropTypes.string.isRequired,
  open: PropTypes.bool.isRequired,
  closeModal: PropTypes.func.isRequired,
};

export { ModalWrapper };
