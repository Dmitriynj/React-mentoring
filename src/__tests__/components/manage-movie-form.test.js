import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { ManageMovieForm } from '../../components/shared/ManageMovieForm';

describe('ManageMovie', () => {
  const initialMovie = {
    title: 'some title',
    release_date: '2021-01-07',
    poster_path: 'some poster path',
    genres: ['Documentary', 'Comedy'],
    runtime: 123,
    overview: 'some text',
  };
  let onConfirmMock;
  let tree;

  beforeEach(() => {
    onConfirmMock = jest.fn();
    tree = render(
      <MemoryRouter initialEntries={['/movies']}>
        <Route exact path="/movies">
          <ManageMovieForm initialMovie={initialMovie} onConfirm={onConfirmMock} />
        </Route>
      </MemoryRouter>
    );
  });

  it('renders correctly', () => {
    expect(tree.container.firstChild).toMatchSnapshot();
    expect(tree.getByText(/title/i)).toBeInTheDocument();
    expect(tree.getByText(/release date/i)).toBeInTheDocument();
    expect(tree.getByText(/submit/i)).toBeInTheDocument();
  });

  it('should pass validations on submit', async () => {
    const rightClick = { button: 2 };

    fireEvent.click(tree.getByText('Submit'), rightClick);
    screen.debug();
    await waitFor(() => expect(onConfirmMock).toHaveBeenCalledWith(initialMovie));
  });
});
