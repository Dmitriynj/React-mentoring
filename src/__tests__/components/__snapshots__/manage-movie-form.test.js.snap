// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageMovie renders correctly 1`] = `
<form>
  <label
    class="MuiFormLabel-root MuiInputLabel-root makeStyles-label-3 MuiInputLabel-animated"
  >
    Title
  </label>
  <div
    class="MuiFormControl-root MuiTextField-root makeStyles-textField-2 MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        class="MuiInputBase-input MuiInput-input"
        id="title"
        name="title"
        type="text"
        value="some title"
      />
    </div>
  </div>
  <label
    class="MuiFormLabel-root MuiInputLabel-root makeStyles-label-3 MuiInputLabel-animated"
  >
    Release date
  </label>
  <div
    class="MuiFormControl-root MuiTextField-root makeStyles-textField-2 MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        class="MuiInputBase-input MuiInput-input"
        id="release_date"
        name="release_date"
        type="date"
        value="2021-01-07"
      />
    </div>
  </div>
  <label
    class="MuiFormLabel-root MuiInputLabel-root makeStyles-label-3 MuiInputLabel-animated"
  >
    Poster path
  </label>
  <div
    class="MuiFormControl-root MuiTextField-root makeStyles-textField-2 MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        class="MuiInputBase-input MuiInput-input"
        id="poster_path"
        name="poster_path"
        type="text"
        value="some poster path"
      />
    </div>
  </div>
  <label
    class="MuiFormLabel-root MuiInputLabel-root makeStyles-label-3 MuiInputLabel-animated"
  >
    Genres
  </label>
  <div
    class="MuiFormControl-root MuiTextField-root makeStyles-textField-2 MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="genres"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="genres"
        role="button"
        tabindex="0"
      >
        Comedy, Documentary
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        name="genres"
        tabindex="-1"
        value="Documentary,Comedy"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <label
    class="MuiFormLabel-root MuiInputLabel-root makeStyles-label-3 MuiInputLabel-animated"
  >
    Runtime
  </label>
  <div
    class="MuiFormControl-root MuiTextField-root makeStyles-textField-2 MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        class="MuiInputBase-input MuiInput-input"
        id="runtime"
        name="runtime"
        type="number"
        value="123"
      />
    </div>
  </div>
  <label
    class="MuiFormLabel-root MuiInputLabel-root makeStyles-label-3 MuiInputLabel-animated"
  >
    Overview
  </label>
  <div
    class="MuiFormControl-root MuiTextField-root makeStyles-textField-2 MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        class="MuiInputBase-input MuiInput-input"
        id="overview"
        name="overview"
        type="text"
        value="some text"
      />
    </div>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-fullWidth"
    tabindex="0"
    type="submit"
  >
    <span
      class="MuiButton-label"
    >
      Submit
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</form>
`;
